File: JavaScript/Ex1.js
// PALETA DE CORES
const ui = {
    reset: "\x1b[0m",
    bold: "\x1b[1m",
    pink: "\x1b[35m",  
    purple: "\x1b[95m",  
    white: "\x1b[97m"      
};

function header(titulo) {
    console.log("\n" + ui.pink + ui.bold + "╭──────────────────────────────────╮");
    console.log(`│  ${titulo.padEnd(30)} │`);
    console.log("╰──────────────────────────────────╯" + ui.reset);
}

// PRODUTOS ABSTRATOS
class VeiculoIndividual {
    iniciarRota() {
        throw new Error("iniciarRota() deve ser implementado");
    }
}

class VeiculoColetivo {
    iniciarRota() {
        throw new Error("iniciarRota() deve ser implementado");
    }
}

// PRODUTO CONCRETO - TERRESTRE
class Carro extends VeiculoIndividual {
    iniciarRota() {
        return "🚗 Carro iniciando trajeto urbano";
    }
}

class Onibus extends VeiculoColetivo {
    iniciarRota() {
        return "🚌 Ônibus transportando passageiros";
    }
}

// PRODUTO CONCRETO - AÉREO
class Helicoptero extends VeiculoIndividual {
    iniciarRota() {
        return "🚁 Helicóptero decolando";
    }
}

class Aviao extends VeiculoColetivo {
    iniciarRota() {
        return "✈️ Avião iniciando voo comercial";
    }
}

// FÁBRICA ABSTRATA
class FabricaTransporte {
    criarVeiculoIndividual() {
        throw new Error("criarVeiculoIndividual() deve ser implementado");
    }

    criarVeiculoColetivo() {
        throw new Error("criarVeiculoColetivo() deve ser implementado");
    }
}

// FÁBRICAS CONCRETAS
class FabricaTerrestre extends FabricaTransporte {
    criarVeiculoIndividual() {
        return new Carro();
    }

    criarVeiculoColetivo() {
        return new Onibus();
    }
}

class FabricaAerea extends FabricaTransporte {
    criarVeiculoIndividual() {
        return new Helicoptero();
    }

    criarVeiculoColetivo() {
        return new Aviao();
    }
}

// CLIENTE
function iniciarTransporte(fabrica) {
    const individual = fabrica.criarVeiculoIndividual();
    const coletivo = fabrica.criarVeiculoColetivo();

    console.log(ui.white + "   ➜ " + individual.iniciarRota() + ui.reset);
    console.log(ui.white + "   ➜ " + coletivo.iniciarRota() + ui.reset);
    console.log(ui.purple + "   ────────────────────────────────" + ui.reset);
}

// USO
header("🌸 TRANSPORTE TERRESTRE");
iniciarTransporte(new FabricaTerrestre());

header("🌸 TRANSPORTE AÉREO");
iniciarTransporte(new FabricaAerea());

File: python/Ex1.py
# PALETA DE CORES
class UI:
    reset = "\033[0m"
    bold = "\033[1m"
    pink = "\033[35m"
    purple = "\033[95m"
    white = "\033[97m"

ui = UI()

def header(titulo):
    print("\n" + ui.pink + ui.bold + "╭──────────────────────────────────╮")
    print(f"│  {titulo.ljust(30)} │")
    print("╰──────────────────────────────────╯" + ui.reset)

# PRODUTOS ABSTRATOS
class VeiculoIndividual:
    def iniciar_rota(self):
        raise NotImplementedError("iniciar_rota() deve ser implementado")

class VeiculoColetivo:
    def iniciar_rota(self):
        raise NotImplementedError("iniciar_rota() deve ser implementado")

# PRODUTO CONCRETO - TERRESTRE
class Carro(VeiculoIndividual):
    def iniciar_rota(self):
        return "🚗 Carro iniciando trajeto urbano"

class Onibus(VeiculoColetivo):
    def iniciar_rota(self):
        return "🚌 Ônibus transportando passageiros"

# PRODUTO CONCRETO - AÉREO
class Helicoptero(VeiculoIndividual):
    def iniciar_rota(self):
        return "🚁 Helicóptero decolando"

class Aviao(VeiculoColetivo):
    def iniciar_rota(self):
        return "✈️ Avião iniciando voo comercial"

# FÁBRICA ABSTRATA
class FabricaTransporte:
    def criar_veiculo_individual(self):
        raise NotImplementedError("criar_veiculo_individual() deve ser implementado")

    def criar_veiculo_coletivo(self):
        raise NotImplementedError("criar_veiculo_coletivo() deve ser implementado")

# FÁBRICAS CONCRETAS
class FabricaTerrestre(FabricaTransporte):
    def criar_veiculo_individual(self):
        return Carro()

    def criar_veiculo_coletivo(self):
        return Onibus()

class FabricaAerea(FabricaTransporte):
    def criar_veiculo_individual(self):
        return Helicoptero()

    def criar_veiculo_coletivo(self):
        return Aviao()

# CLIENTE
def iniciar_transporte(fabrica):
    individual = fabrica.criar_veiculo_individual()
    coletivo = fabrica.criar_veiculo_coletivo()

    print(ui.white + "   ➜ " + individual.iniciar_rota() + ui.reset)
    print(ui.white + "   ➜ " + coletivo.iniciar_rota() + ui.reset)
    print(ui.purple + " ────────────────────────────────" + ui.reset)

# USO
header("🌸 TRANSPORTE TERRESTRE")
iniciar_transporte(FabricaTerrestre())

header("🌸 TRANSPORTE AÉREO")
iniciar_transporte(FabricaAerea())